FluidNC/src/Spindles/HBridgeSpindle.cpp

```diff
@@ -0,0 +1,195 @@
+        setupSpeeds(_pwm_freq);
+
+        if (_output_cw_pin.defined()) {
+            if (_output_cw_pin.capabilities().has(Pin::Capabilities::PWM)) {
+                auto outputNative = _output_cw_pin.getNative(Pin::Capabilities::PWM);
+                _pwm_cw_chan_num  = ledcInit(_output_cw_pin, -1, (double)_pwm_freq, _pwm_precision);
+            } else {
+                log_error(name() << " output_cw_pin " << _output_cw_pin.name().c_str() << " cannot do PWM");
+            }
+        } else {
+            log_error(name() << " output_cw_pin not defined");
+        }
+
+        if (_output_ccw_pin.defined()) {
+            if (_output_ccw_pin.capabilities().has(Pin::Capabilities::PWM)) {
+                auto outputBNative = _output_ccw_pin.getNative(Pin::Capabilities::PWM);
+                _pwm_ccw_chan_num  = ledcInit(_output_ccw_pin, -1, (double)_pwm_freq, _pwm_precision);
+            } else {
+                log_error(name() << " output_ccw_pin " << _output_ccw_pin.name().c_str() << " cannot do PWM");
+            }
+        } else {
+            log_error(name() << " output_ccw_pin not defined");
+        }
+
+        _current_state    = SpindleState::Disable;
+        _current_pwm_duty = 0;
+        _enable_pin.setAttr(Pin::Attr::Output);
+
+        if (_speeds.size() == 0) {
+            // The default speed map for a PWM spindle is linear from 0=0% to 10000=100%
+            linearSpeeds(10000, 100.0f);
+        }
+        setupSpeeds(_pwm_period);
+        config_message();
+    }
+
+    // Get the GPIO from the machine definition
+    void HBridgeSpindle::get_pins_and_settings() {
+        // setup all the pins
+
+        is_reversable = _output_ccw_pin.defined();
+
+        _pwm_precision = calc_pwm_precision(_pwm_freq);  // determine the best precision
+        _pwm_period    = (1 << _pwm_precision);
+    }
+
+    void IRAM_ATTR HBridgeSpindle::set_enable(bool enable) {
+        if (_disable_with_zero_speed && sys.spindle_speed == 0) {
+            enable = false;
+        }
+
+        _enable_pin.synchronousWrite(enable);
+    }
+
+    void IRAM_ATTR HBridgeSpindle::setSpeedfromISR(uint32_t dev_speed) {
+        set_enable(gc_state.modal.spindle != SpindleState::Disable);
+        set_output(dev_speed);
+    }
+
+    void HBridgeSpindle::setState(SpindleState state, SpindleSpeed speed) {
+        if (sys.abort) {
+            return;  // Block during abort.
+        }
+
+        if (!_output_cw_pin.defined() || !_output_ccw_pin.defined()) {
+            log_warn(name() << " spindle pins not defined");
+        }
+
+        // We always use mapSpeed() with the unmodified input speed so it sets
+        // sys.spindle_speed correctly.
+        uint32_t dev_speed = mapSpeed(speed);
+        if (state == SpindleState::Disable) {  // Halt or set spindle direction and speed.
+            dev_speed = 0;
+        } else {
+            // PWM: this could wreak havoc if the direction is changed without first
```

Comment:
What information does the "PWM:" part of this line convey?  Actually, the whole comment could probably be eliminated because it is the core code's responsibility to take care of direction changes and this module is not different from any other spindle in this respect.